src/components/lab/EmittedLine.js

```diff
@@ -1,168 +1,74 @@
+  MOLECULE_CENTER_Y_FROM_BOTTOM_OF_CANVAS,
   NITROUS_OXIDE_MOLECULE_ID,
-  CARBON_DIOXIDE_MOLECULE_LOWEST_Y,
-  OZONE_MOLECULE_LOWEST_Y,
-  METHANE_MOLECULE_LOWEST_Y,
-  WATER_MOLECULE_LOWEST_Y,
-  NITROUS_OXIDE_MOLECULE_LOWEST_Y,
+  OZONE_MOLECULE_ID,
+  SINE_CURVE_AMPLITUDE,
+  SPECTRUMS,
+  VISIBLE_LIGHT_PERIOD,
+  WATER_MOLECULE_ID,
+  Y_INCREMENT_PER_POINT,
+  Y_SHIFT_PER_INTERVAL,
 } from '../../config/constants';
-import { toggleMoleculeOscillation, updateLinePoints } from '../../actions';
+import generateSineCurve from '../../utils/generateSineCurve';
 
-class EmittedLine extends Component {
-  static propTypes = {
-    x: PropTypes.number.isRequired,
-    stageDimensions: PropTypes.shape({
-      height: PropTypes.number,
-      width: PropTypes.number,
-    }).isRequired,
-    isPaused: PropTypes.bool.isRequired,
-    lineIndex: PropTypes.number.isRequired,
-    emittedLines: PropTypes.arrayOf(
-      PropTypes.shape({
-        points: PropTypes.arrayOf(PropTypes.number),
-        oscillationConstant: PropTypes.number.isRequired,
-      }),
-    ).isRequired,
-    dispatchUpdateLinePoints: PropTypes.func.isRequired,
-    spectrum: PropTypes.string.isRequired,
-    moleculesOnCanvas: PropTypes.arrayOf(
-      PropTypes.shape({
-        molecule: PropTypes.string.isRequired,
-        moleculeAreaStatus: PropTypes.string.isRequired,
-        shouldOscillate: PropTypes.bool.isRequired,
-      }),
-    ).isRequired,
-    dispatchToggleMoleculeOscillation: PropTypes.func.isRequired,
-  };
+const EmittedLine = ({ x, lineIndex }) => {
+  const intervalCount = useSelector(({ lab }) => lab.intervalCount);
+  const moleculesOnCanvas = useSelector(({ lab }) => lab.moleculesOnCanvas);
+  const spectrum = useSelector(({ lab }) => lab.spectrum);
+  const { height: stageHeight } = useSelector(
+    ({ layout }) => layout.lab.stageDimensions,
+  );
+  const currentLineMolecule = moleculesOnCanvas[lineIndex].molecule;
 
-  componentDidUpdate({ isPaused: prevIsPaused }) {
-    const { isPaused } = this.props;
-    if (isPaused !== prevIsPaused && isPaused) {
-      clearInterval(this.emittedLineInterval);
-    } else if (isPaused !== prevIsPaused && !isPaused) {
-      this.beginLineInterval();
+  let period = VISIBLE_LIGHT_PERIOD;
+  // Ozone, Methane, Water, CO2, and Nitrous Oxide are greenhouse gases, and hence absorb infrared radiation
+  // By default, the radiation 'absorption point' is 0 (i.e. the top of the canvas), i.e. the radiation line extends to the top of the canvas
+  // With a greenhouse molecule, the absorption point is the center of the molecule
+  let absorptionPoint = 0;
+  if (spectrum === SPECTRUMS.INFRARED) {
+    period = INFRARED_RADIATION_PERIOD;
+    if (
+      currentLineMolecule === OZONE_MOLECULE_ID ||
+      currentLineMolecule === METHANE_MOLECULE_ID ||
+      currentLineMolecule === WATER_MOLECULE_ID ||
+      currentLineMolecule === CARBON_DIOXIDE_MOLECULE_ID ||
+      currentLineMolecule === NITROUS_OXIDE_MOLECULE_ID
```

Comment:
I would create a constant arrays `ABSORBING_MOLECULE_IDS` and check if the id is contained.

Reply:
nice one